README.md
# Draw Me A Kanji [![Build Status](https://travis-ci.org/mbilbille/dmak.png?branch=master)](https://travis-ci.org/mbilbille/dmak) [![Bitdeli Badge](https://d2weczhvl823v0.cloudfront.net/mbilbille/dmak/trend.png)](https://bitdeli.com/free "Bitdeli Badge")

> A funny drawer for all your Japanese writings

![Draw Me A Kanji](https://raw.github.com/mbilbille/dmak/gh-pages/images/sample.png)

*For the sake of simplicity, "Draw Me A Kanji" is shortened to dmak*

## Usage
Using dmak.js is easy, a lot easier than writing kanjis, trust me!

**Prerequisite**
To simplify all the work with SVG, we use the cool JavaScript library *Raphaël* ➲ [Download](http://github.com/DmitryBaranovskiy/raphael/raw/master/raphael-min.js)
*and no worries this is the ONLY dependency.*

Once you are all set, include both `raphael.js` and `dmak.js` files in your HTML file.

Up to you to use either:

#### Pure JavaScript
```html
<div id="draw"></div>
<script>
  var dmak = new Dmak('電車', {
    'element' : "draw"
  });
</script>
```

#### or the jQuery plugin

```html
<div id="draw"></div>
<script>
    // That's it!
    $("#draw").dmak('電車');
</script>
```
**NOTE:** *You need to include the additional `jquery.dmak.js` file*

## Options
Dmak.js comes with a whole set of options to alter the way it behaves. Here is an explicit list of all parameters available to you followed by their default value.

* `uri` - path to the SVG files folder. `""`
* `skipLoad` - skip SVG files loading. `false`
* `autoplay` - start drawing as soon as all SVG files are loaded. `true`
* `height` - height in pixels of a single paper surface. `109`
* `width` - width in pixels of a single paper surface. `109`
* `viewBox.x` - x position of the canvas. `0`
* `viewBox.y` - y position of the canvas. `0`
* `viewBox.w` - width of the canvas. `109`
* `viewBox.h` - height of the canvas. `109`
* `step` - positive integer which defines the speed of the drawing. `0.03`
* `element` - DOM element or its ID which is going to be a parent for drawing surface. `"draw"`
* `stroke.animated` - enable or disable stroke animation. `true`
* `stroke.order.visible` - display stroke order. `false`
* `stroke.order.visible.attr.font-size` - stroke order font size in pixels. `8`
* `stroke.order.visible.attr.fill` - stroke order color. `#999999`
* `stroke.attr.active` - currently drawn stroke color. `"#BF0000"`
* `stroke.attr.stroke` - stroke color (can use "random" keyword here for random color). `"#2C2C2C"`
* `stroke.attr.stroke-width` - stroke width in pixels. `4`
* `stroke.attr.stroke-linecap` - ["butt", "square", "round"]. `"round`
* `stroke.attr.stroke-linejoin` - ["bevel", "round", "miter"]. `"round`
* `grid.show` - show or hide gridlines. `true`
* `grid.attr.stroke` - grid color. `"#CCCCCC"`
* `grid.attr.stroke-width` - grid width in pixels. `0.5`
* `grid.attr.stroke-dasharray` - ["", "-", ".", "-.", "-..", ". ", "- ", "--", "- .", "--.", "--.."]. `"--"`
* `loaded` - callback function which is executed when all SVG files are fully loaded
* `erased` - callback function which is executed when a stroke is erased
* `drew` - callback function which is executed when a stroke is drawn

## Demo
For basic samples please refer to [demo](https://github.com/mbilbille/dmak/tree/master/demo) folder, otherwise dive into drawmeakanji.com [source code](https://github.com/mbilbille/dmak/tree/gh-pages).

## Compatibility
- Chrome 1.0+
- Firefox 16.0+ (see: [bugzilla.mozilla.org/show_bug.cgi?id=902879](https://bugzilla.mozilla.org/show_bug.cgi?id=902879))
- Opera 17.0+
- Safari 3.0+
- ~~IE~~

## Inspirations

- From @jakearchibald and his great article on [Animated line drawing in SVG](http://jakearchibald.com/2013/animated-line-drawing-svg/)
- From @akeru and his [kanjiviewer.js](https://github.com/KanjiVG/kanjivg.github.com/blob/master/js/kanjiviewer.js) library.

And of course a huge thanks and support to [KanjiVG](http://kanjivg.tagaini.net) for providing a whole set of SVG files.

## Contributing

Check [CONTRIBUTING.md](https://github.com/mbilbille/dmak/tree/master/CONTRIBUTING.md)

## History

Check [Release](https://github.com/mbilbille/dmak/releases) list.

## Building

- Install [nodejs](http://nodejs.org/)
- Install [grunt-cli](http://gruntjs.com/getting-started)
- Process dependencies by running `npm install` in the repository root folder
- Build a new release by invoking `grunt` in the repository root folder

## License

*Draw Me A Kanji* (dmak.js) was created by [Matthieu Bilbille](http://github.com/mbilbille) and released under the [MIT License](http://github.com/mbilbille/dmak/blob/master/LICENSE).


HISTORY.md
# History

* v0.1.0 January 2nd, 2013  
	First release
* v0.0.1 December 20th, 2013  
	Initial commit


Gruntfile.js
module.exports = function(grunt) {

	grunt.initConfig({

		// Import package manifest
		pkg: grunt.file.readJSON("package.json"),

		// Banner definitions
		meta: {
			banner: "/*\n" +
			" *  <%= pkg.title || pkg.name %> - v<%= pkg.version %>\n" +
			" *  <%= pkg.description %>\n" +
			" *  <%= pkg.homepage %>\n" +
			" *\n" +
			" *  Made by <%= pkg.author %>\n" +
			" *  Under <%= pkg.licenses[0].type %> License\n" +
			" */\n"
		},

		// Concat definitions
		concat: {
			dist: {
				files : {
					"dist/dmak.js" : ["src/dmak.js", "src/dmakLoader.js"],
					"dist/jquery.dmak.js" : ["src/jquery.dmak.js"],
				}
			},
			options: {
				banner: "<%= meta.banner %>",
			}
		},

		// Lint definitions
		jshint: {
			files: ["src/dmakLoader.js", "src/dmak.js", "src/jquery.dmak.js"],
			options: {
				jshintrc: ".jshintrc"
			}
		},

		// Minify definitions
		uglify: {
			my_target: {
				files: {
					"dist/dmak.min.js" : "dist/dmak.js",
					"dist/jquery.dmak.min.js" : "dist/jquery.dmak.js",
				}
			},
			options: {
				banner: "<%= meta.banner %>",
				report: 'gzip'
			}
		}
	});

	grunt.loadNpmTasks("grunt-contrib-concat");
	grunt.loadNpmTasks("grunt-contrib-jshint");
	grunt.loadNpmTasks("grunt-contrib-uglify");

	grunt.registerTask("default", ["jshint", "concat", "uglify"]);
	grunt.registerTask("travis", ["jshint"]);
};


src/dmakLoader.js
;(function () {

	"use strict";

	// Create a safe reference to the DrawMeAKanji object for use below.
	var DmakLoader = function (uri) {
		this.uri = uri;
	};

	/**
	 * Gather SVG data information for a given set of characters.
	 * By default this action is done while instanciating the Word
	 * object, but it can be skipped, see above
	 */
	DmakLoader.prototype.load = function (text, callback) {
		var paths = [],
			nbChar = text.length,
			done = 0,
			i,
			callbacks = {
				done: function (index, data) {
					paths[index] = data;
					done++;
					if (done === nbChar) {
						callback(paths);
					}
				},
				error: function (msg) {
					console.log("Error", msg);
				}
			};

		for (i = 0; i < nbChar; i++) {
			loadSvg(this.uri, i, text.charCodeAt(i).toString(16), callbacks);
		}
	};

	/**
	 * Try to load a SVG file matching the given char code.
	 * @thanks to the incredible work made by KanjiVG
	 * @see: http://kanjivg.tagaini.net
	 */
	function loadSvg(uri, index, charCode, callbacks) {
		var xhr = new XMLHttpRequest(),
			code = ("00000" + charCode).slice(-5);

		// Skip space character
		if(code === "00020" || code === "03000") {
			callbacks.done(index, {
				paths: [],
				texts: []
			});
			return;
		}

		xhr.open("GET", uri + code + ".svg", true);
		xhr.onreadystatechange = function () {
			if (xhr.readyState === 4) {
				if (xhr.status === 200) {
					callbacks.done(index, parseResponse(xhr.response));
				} else {
					callbacks.error(xhr.statusText);
				}
			}
		};
		xhr.send();
	}

	/**
	 * Simple parser to extract paths and texts data.
	 */
	function parseResponse(response) {
		var data = {
				paths: [],
				texts: []
			},
			dom = new DOMParser().parseFromString(response, "application/xml"),
			paths = dom.querySelectorAll("path"),
			texts = dom.querySelectorAll("text"),
			i;

		for (i = 0; i < paths.length; i++) {
			data.paths.push(paths[i].getAttribute("d"));
		}

		for (i = 0; i < texts.length; i++) {
			data.texts.push({
				"value" : texts[i].textContent,
				"x" : texts[i].getAttribute("transform").split(" ")[4],
				"y" : texts[i].getAttribute("transform").split(" ")[5].replace(")", "")
			});
		}
		return data;
	}

	window.DmakLoader = DmakLoader;
}());


js/website.dmak.js

// Add segments to a slider
$.fn.addSliderSegments = function (amount) {
    return this.each(function () {
        var segmentGap = 100 / (amount - 1) + "%"
        , segment = "<div class='ui-slider-segment' style='margin-left: " + segmentGap + ";'></div>";
        $(this).prepend(segment.repeat(amount - 2));
    });
};

// Implement our own repeat method() which copies the current string a given 
// times and returns the new string
String.prototype.repeat = function( num )
{
    return new Array( num + 1 ).join( this );
}

var p;
var options = {
    step : 0.03,
    uri: 'kanji/',
    stroke: {
        order: {
            'visible' : false,
            attr : {
                'fill' : '#CDC3C7'
            }
        },
        attr : {
            'active': '#E74C3C',
            'stroke': '#7F8C8D',
            'stroke-width': 4,
        }
    },
    grid : {
        show: true,
    },
    loaded: function(strokes) {
        $('.logo').toggleClass('flip');
        $(".wrap-draw").addClass("slideDown");
        $slider = $("#slider");
        if ($slider.length > 0) {
            p = 1;
            $slider.slider({
                min: 1,
                max: strokes.length + 1,
                value: 1,
                orientation: "horizontal",
                range: "min",
                slide: function(event, ui) {
                    var range = ui.value - p;
                    var action = 'forward';
                    if(range < 0) {
                        action = 'rewind';
                        range = range * -1;
                    }
                    p = ui.value;

                    for (var i = 0; i < range; i++) {
                        $("#draw").dmak(action, 1);
                    };
                }
            }).addSliderSegments($slider.slider("option").max);
        }
    },
    erased: function(index) {
        if($slider.slider("value") > index) {
            p = index + 1;
            $slider.slider( "value", p );
       }
   },
   drew: function(index) {
        if($slider.slider("value") <= index) {
            p = index + 1;
            $slider.slider( "value", p );
        }
    }
};

$(function() {

    var blankDrawer = $(".wrap-draw").html();

    // Check if we have a word in the URL
    if(window.location.hash) {
        var word = window.location.hash.substring(1);
        // Properly decode UTF8 characters
        word = decodeURIComponent(word);
        ga('send', 'event', 'url', 'search', word);
        $("#draw").dmak(word, options);
    }

    $('#mybtn').click(function(e) {
        e.preventDefault();

        if ($('#draw').data("plugin_dmak")){
            $("#draw").dmak("pause");
            $('.wrap-draw').html(blankDrawer);

        }
        var text = $('#mytext').val();
        window.location.hash = "#" + text;
        ga('send', 'event', 'input', 'search', text);
        $('#mytext').val("");
        $("#draw").dmak(text, options);
    });

    $( document ).on( "click", "#play", function(e){
        e.preventDefault();
       $('#draw').dmak('play');
    });

    $( document ).on( "click", "#pause", function(e){
        e.preventDefault();
        $('#draw').dmak('pause');
    });

    $( document ).on( "click", "#rewind", function(e){
        e.preventDefault();
        $('#draw').dmak('rewind', 1);
    });

    $( document ).on( "click", "#forward", function(e){
        e.preventDefault();
        $('#draw').dmak('forward', 1);
    });

    $( document ).on( "click", "#sample-words span", function(e){
        e.preventDefault();
        $('#mytext').val($(this).html());
    });


    // SETTINGS
    var $sliderWidth = $("#dmak-stroke-width");
    $sliderWidth.slider({
        min: 1,
        max: 9,
        value: options.stroke.attr['stroke-width'],
        orientation: "horizontal",
        range: "min",
        stop: function(event, ui) {
            options.stroke.attr['stroke-width'] = ui.value;
        }
    }).addSliderSegments($sliderWidth.slider("option").max);

    $("#dmak-color").val(options.stroke.attr.stroke);
    $( document ).on( "change", "#dmak-color", function(e){
        options.stroke.attr.stroke = $(this).val();
    });

    $("#dmak-color-active").val(options.stroke.attr.active);
    $( document ).on( "change", "#dmak-color-active", function(e){
        options.stroke.attr.active = $(this).val();
    });

    var speeds = [0.1, 0.05, 0.02, 0.009, 0.005]
    var $sliderSpeed = $("#dmak-slider-speed");
    $sliderSpeed.slider({
        min: 1,
        max: 5,
        value: 3,
        orientation: "horizontal",
        range: "min",
        stop: function(event, ui) {
            options.step = speeds[ui.value - 1];
        }
    }).addSliderSegments($sliderSpeed.slider("option").max);

    $("#dmak-order").html(options.stroke.order.visible ? 'ON' : 'OFF');
    $( document ).on( "click", "#dmak-order", function(e){
        e.preventDefault();
        if($(this).html() == "ON") {
            options.stroke.order.visible = false;
            $(this).html("OFF");
        } else {
            options.stroke.order.visible = true;
            $(this).html("ON");
        }
    });

    $("#dmak-grid").html(options.grid.show ? 'ON' : 'OFF');
    $( document ).on( "click", "#dmak-grid", function(e){
        e.preventDefault();
        if($(this).html() == "ON") {
            options.grid.show = false;
            $(this).html("OFF");
        } else {
            options.grid.show = true;
            $(this).html("ON");
        }
    });
});


src/jquery.dmak.js
/**
 * Draw Me a Kanji - version: 0.1
 * Plugin jQuery
 *
 * https://github.com/mbilbille/drawmeakanji
 * https://github.com/mbilbille
 *
 * Licensed under the MIT license.
 */
 ;(function ( $, window, document, undefined ) {

	// Create the defaults once
	var pluginName = "dmak";

	// The actual plugin constructor
	function Plugin ( element, text, options ) {
		this.element = element;
		this.text = text;
		this.options = $.extend( {}, options, {"element" : $(element).attr("id")} );
		this._name = pluginName;
		this.dmak = null;
		this.init();
	}

	Plugin.prototype = {
		init: function () {
			this.dmak = new Dmak(this.text, this.options);
		},
		reset: function() {
			this.dmak.erase();
		},
		pause: function() {
			this.dmak.pause();
		},
		play: function() {
			this.dmak.render();
		},
		rewind: function (x) {
			this.dmak.eraseLastStrokes(x);
		},
		forward: function (x) {
			this.dmak.renderNextStrokes(x);
		},
		rewindTo: function (t) {
			this.dmak.erase(t);
		},
		forwardTo: function (t) {
			this.dmak.render(t);
		}
	};

	// A really lightweight plugin wrapper around the constructor,
	// preventing against multiple instantiations
	$.fn[ pluginName ] = function ( text, options ) {
		return this.each(function() {
			if ( !$.data( this, "plugin_" + pluginName ) ) {
				$.data( this, "plugin_" + pluginName, new Plugin( this, text, options ) );
			}
			else if ($.isFunction(Plugin.prototype[text])) {
				$.data(this, "plugin_" + pluginName)[text](options);
			}
		});
	};

})( jQuery, window, document );


src/dmak.js
 ;(function () {

	"use strict";

	// Create a safe reference to the DrawMeAKanji object for use below.
	var Dmak = function (text, options) {
		this.text = text;
		this.options = extend(Dmak.options, options);
		this.strokes = [];
		this.papers = [];
		this.pointer = 0;
		this.timeouts = [];

		if (!this.options.skipLoad) {
			var loader = new DmakLoader(this.options.uri);
			loader.load(text, function (data) {
				this.prepare(data);

				// Execute custom callback "loaded" here
				this.options.loaded(this.kanjis);

				if (this.options.autoplay) {
					this.render();
				}
			}.bind(this));
		}
	};

	// Current version.
	Dmak.VERSION = "0.2.0";

	Dmak.options = {
		uri: "",
		skipLoad: false,
		autoplay: true,
		height: 109,
		width: 109,
		viewBox: {
			x: 0,
			y: 0,
			w: 109,
			h: 109
		},
		step: 0.03,
		element: "draw",
		stroke: {
			animated: true,
			order: {
				visible: false,
				attr: {
					"font-size": "8",
					"fill": "#999999"
				}
			},
			attr: {
				"active": "#BF0000",
				// may use the keyword "random" here for random color
				"stroke": "#2C2C2C",
				"stroke-width": 4,
				"stroke-linecap": "round",
				"stroke-linejoin": "round"
			}
		},
		grid: {
			show: true,
			attr: {
				"stroke": "#CCCCCC",
				"stroke-width": 0.5,
				"stroke-dasharray": "--"
			}
		},
		loaded: function () {
		},
		erased: function () {
		},
		drew: function () {
		}
	};

	Dmak.fn = Dmak.prototype = {

		/**
		 * Prepare kanjis and papers for rendering.
		 */
		prepare: function (data) {
			this.kanjis = preprocessStrokes(data);
			this.papers = giveBirthToRaphael(data.length);
			if (this.options.grid.show) {
				showGrid(this.papers);
			}
		},

		/**
		 * Clean all strokes on papers.
		 */
		erase: function (end) {
			// Cannot have two rendering process for
			// the same draw. Keep it cool.
			if (this.timeouts.length) {
				return false;
			}

			// Don't go behind the beginning.
			if (this.pointer <= 0) {
				return false;
			}

			if (typeof end === "undefined") {
				end = 0;
			}

			do {
				this.pointer--;

				// In some cases the text object may be null:
				//  - Stroke order display disabled
				//  - Stroke already deleted
				if (this.kanjis[this.pointer].object.text !== null) {
					this.kanjis[this.pointer].object.text.remove();
				}
				// No worries path can not be null here
				this.kanjis[this.pointer].object.path.remove();

				// Finally properly prepare the object variable
				this.kanjis[this.pointer].object = {
					"path" : null,
					"text" : null
				};

				// Execute custom callback "erased" here
				this.options.erased(this.pointer);
			}
			while (this.pointer > end);
		},

		/**
		 * All the magic happens here.
		 */
		render: function (end) {

			// Cannot have two rendering process for
			// the same draw. Keep it cool.
			if (this.timeouts.length) {
				return false;
			}

			if (typeof end === "undefined") {
                end = this.kanjis.length;
            } else if (end > this.kanjis.length) {
				return false;
			}

			var cb = function (that) {
					createStroke(that.papers[that.kanjis[that.pointer].char], that.kanjis[that.pointer]);
					that.pointer++;
					that.timeouts.shift();

					// Execute custom callback "drew" here
					that.options.drew(that.pointer);
				},
				delay = 0,
				i,
				t;

			for (i = this.pointer; i < end; i++) {
				if (!Dmak.options.stroke.animated || delay <= 0) {
					cb(this);
				} else {
					t = setTimeout(cb, delay, this);
					this.timeouts.push(t);
				}
				delay += this.kanjis[i].duration;
			}
		},

		/**
		 * Pause rendering
		 */
		pause: function () {
			for (var i = 0; i < this.timeouts.length; i++) {
				window.clearTimeout(this.timeouts[i]);
			}
			this.timeouts = [];
		},

		/**
		 * Wrap the erase function to remove the x last strokes.
		 */
		eraseLastStrokes: function (nbStrokes) {
			this.erase(this.pointer - nbStrokes);
		},

		/**
		 * Wrap the render function to render the x next strokes.
		 */
		renderNextStrokes: function (nbStrokes) {
			this.render(this.pointer + nbStrokes);
		}

	};

	// HELPERS

	/**
	 * Flattens the array of strokes ; 3D > 2D and does some preprocessing while
	 * looping through all the strokes:
	 *  - Maps to a character index
	 *  - Calculates path length
	 */
	function preprocessStrokes(data) {
		var strokes = [],
			stroke,
			length,
			i,
			j;

		for (i = 0; i < data.length; i++) {
			for (j = 0; j < data[i].paths.length; j++) {
				length = Raphael.getTotalLength(data[i].paths[j]);
				stroke = {
					"char": i,
					"length": length,
					"duration": length * Dmak.options.step * 1000,
					"path": data[i].paths[j],
					"text": data[i].texts[j],
					"object": {
						"path" : null,
						"text": null
					}
				};
				strokes.push(stroke);
			}
		}

		return strokes;
	}

	/**
	 * Init Raphael paper objects
	 */
	function giveBirthToRaphael(nbChar) {
		var papers = [],
			paper,
			i;

		for (i = 0; i < nbChar; i++) {
			paper = new Raphael(Dmak.options.element, Dmak.options.width + "px", Dmak.options.height + "px");
			paper.setViewBox(Dmak.options.viewBox.x, Dmak.options.viewBox.y, Dmak.options.viewBox.w, Dmak.options.viewBox.h);
			paper.canvas.setAttribute("class", "dmak-svg");
			papers.push(paper);
		}
		return papers.reverse();
	}

	/**
	 * Draw the background grid
	 */
	function showGrid(papers) {
		var i;

		for (i = 0; i < papers.length; i++) {
			papers[i].path("M" + (Dmak.options.viewBox.w / 2) + ",0 L" + (Dmak.options.viewBox.w / 2) + "," + Dmak.options.viewBox.h).attr(Dmak.options.grid.attr);
			papers[i].path("M0," + (Dmak.options.viewBox.h / 2) + " L" + Dmak.options.viewBox.w + "," + (Dmak.options.viewBox.h / 2)).attr(Dmak.options.grid.attr);
		}
	}

	/**
	 * Draw a single stroke ; drawing can be animated if set as so.
	 */
	function createStroke(paper, stroke) {
		stroke.object.path = paper.path(stroke.path);
		stroke.object.path.attr(Dmak.options.stroke.attr);

		if (Dmak.options.stroke.animated) {
			animateStroke(paper, stroke);
		}
		else if(Dmak.options.stroke.attr.stroke === "random") {
			// Get random color if set so and stroke animation is disable
			stroke.object.path.node.style.stroke = Raphael.getColor();
		}
	}

	/**
	 * Draw a single next to
	 */
	function showStrokeOrder(paper, stroke) {
		stroke.object.text = paper.text(stroke.text.x, stroke.text.y, stroke.text.value);
		stroke.object.text.attr(Dmak.options.stroke.order.attr);
	}

	/**
	 * Animate stroke drawing.
	 * Based on the great article wrote by Jake Archibald
	 * http://jakearchibald.com/2013/animated-line-drawing-svg/
	 */
	function animateStroke(paper, stroke) {
		stroke.object.path.attr({"stroke": Dmak.options.stroke.attr.active});
		stroke.object.path.node.style.transition = stroke.object.path.node.style.WebkitTransition = "none";

		// Set up the starting positions
		stroke.object.path.node.style.strokeDasharray = stroke.length + " " + stroke.length;
		stroke.object.path.node.style.strokeDashoffset = stroke.length;

		// Trigger a layout so styles are calculated & the browser
		// picks up the starting position before animating
		stroke.object.path.node.getBoundingClientRect();
		stroke.object.path.node.style.transition = stroke.object.path.node.style.WebkitTransition = "stroke-dashoffset " + stroke.duration + "ms ease";

		// Go!
		stroke.object.path.node.style.strokeDashoffset = "0";

		// Revert back to the options color when the animation is done.
		setTimeout(function () {
			// The stroke object may have been already erased when we reach this timeout
			if (stroke.object.path === null) {
				return;
			}

			if (Dmak.options.stroke.order.visible) {
				showStrokeOrder(paper, stroke);
			}

			var color = Dmak.options.stroke.attr.stroke;
			if(Dmak.options.stroke.attr.stroke === "random") {
				color = Raphael.getColor();
			}

			stroke.object.path.node.style.stroke = color;
			stroke.object.path.node.style.transition = stroke.object.path.node.style.WebkitTransition = "stroke 400ms ease";
		}, stroke.duration);
	}

	/**
	 * Simplistic helper function for extending objects
	 */
	function extend(defaults, replacement) {
		var result = defaults,
			key;

		if (arguments.length !== 2) {
			throw new Error("Missing arguments in extend function");
		}

		for (key in replacement) {
			if (typeof result[key] === "object") {
				result[key] = extend(result[key], replacement[key]);
			} else if (result.hasOwnProperty(key)) {
				result[key] = replacement[key];
			}
		}
		return result;
	}

	window.Dmak = Dmak;
}());
